### ISSUE TITLE
Pipeline errors if a suitable worker is not available at start

### ISSUE THREAD

**Issue by kberzinch** [concourse/concourse#7566]
## Summary

In 7.4.0, if a suitable worker was not available when the job started, it would wait for a worker to be available:

```
no suitable workers found, waiting for worker...
```
![7 4 0](https://user-images.githubusercontent.com/10482109/133933810-4f2474da-e9e5-430d-b7c9-e291136b4b18.png)

The same situation now causes the job to immediately error in 7.5.0:

```
no workers satisfying: resource type 'git', tag 'some-worker-tag', version: '2.3'
```
![7 5 0](https://user-images.githubusercontent.com/10482109/133933841-355c9544-c509-4fab-b4bc-463e98d247e4.png)

## Steps to reproduce

Example pipeline to reproduce the issue:

```yaml
---
resources:

- name: concourse
  type: git
  public: true
  icon: concourse
  source:
    branch: master
    uri: https://github.com/concourse/concourse

jobs:

- name: get-concourse
  plan:
  - get: concourse
    tags:
    - some-worker-tag
```

Start up Concourse 7.4.0:

```sh
# Get the starter Docker Compose file
$ wget https://concourse-ci.org/docker-compose.yml

# Set the tag for the concourse/concourse image to 7.4
$ sed -i 's/concourse\/concourse/concourse\/concourse:7.4/g' docker-compose.yml

# Bring up Concourse
$ docker-compose up
```

Run the job:

```sh
# Assuming target is already configured and logged in
$ fly --target local set-pipeline --pipeline demo --config pipeline.yml
$ fly --target local unpause-pipeline --pipeline demo
$ fly --target local trigger-job --job demo/get-concourse
```

At this point the build is running and waiting for a worker as shown above. Abort the job.

Start up Concourse 7.5.0:

```sh
# Stop existing Concourse stack
$ docker-compose down

# Set the tag for the concourse/concourse image to 7.5
$ sed -i 's/concourse\/concourse:7.4/concourse\/concourse:7.5/g' docker-compose.yml

# Bring up Concourse
$ docker-compose up
```

Run the job again:

```sh
# Assuming target is still configured and logged in, and fly is synced
$ fly --target local trigger-job --job demo/get-concourse
```

## Expected results

The job waits for an available worker.

## Actual results

The job errors immediately as shown above.

## Additional context

I was using this behavior for cost optimization - stop all workers when no jobs are running, start again when `concourse_steps_waiting` metric is greater than 0. (It's a little more complicated than that, but there's the short version.)

I did notice some possibly related issues/PRs which sound like this change might be intentional, but I figured I'd open an issue to clarify - #7467, #6597

## Triaging info

* Concourse version: 7.5.0
* Browser (if applicable): Chrome 93 on Ubuntu 20.04
* Did this used to work? Yes, in 7.4.0

**Comment by clarafu** [comment]
@kberzinch That's a really interesting use case with stopping and starting the workers only when jobs are actually running. Unfortunately you're right that the change was intentional with #6597. I will make a change to the release note to make that more explicit. We made this change because previously, if the worker that was needed does not exist it will hang the job forever. In regular scenarios, a user might have declared a job with a tag but the tagged worker does not exist, in which case the job will just hang forever waiting for the tagged worker to appear.

That being said, I think there is a valid argument for keeping it waiting for a suitable worker to appear. I think in general it might be better use experience because of the fact that the pre-7.5.0 behaviour was having it wait for a worker and also if the suitable worker disconnects for a bit and re-appears, the build will just wait for the worker to come back rather than erroring. 

I'd like to ping @aoldershaw to ask for his opinion on this (if he is around). We can also have some further discussions because I am definitely open to reverting back to the old behaviour if that is the solution we come up with. Possibly just have it wait and then timeout after a period of time to avoid stuck builds waiting on workers that will not come back?

**Comment by kberzinch** [comment]
👍🏻 Thanks for clarifying. I do understand the intent to fail early and loud, and I suspect this is a bit of a corner case as far as Concourse deployments go. It's definitely not a dealbreaker for me.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
